garak/generators/litellm.py

```diff
@@ -0,0 +1,159 @@
+"""LiteLLM model support
+
+Support for LiteLLM, which allows calling LLM APIs using the OpenAI format.
+
+Depending on the model name provider, LiteLLM automatically
+reads API keys from the respective environment variables.
+(e.g. OPENAI_API_KEY for OpenAI models)
+
+API key can also be directly set in the supplied generator json config.
+This also enables support for any custom provider that follows the OAI format.
+
+e.g Supply a JSON like this for Ollama's OAI api:
+```json
+{
+    "litellm.LiteLLMGenerator" : {
+        "api_base" : "http://localhost:11434/v1",
+        "provider" : "openai",
+        "api_key" : "test"
+    }
+}
+```
+"""
+
+import logging
+
+from os import getenv
+from typing import List, Union
+
+import backoff
+
+import litellm
+
+from garak import _config
+from garak.generators.base import Generator
+
+# Fix issue with Ollama which does not support `presence_penalty`
+litellm.drop_params = True
+# Suppress log messages from LiteLLM
+litellm.verbose_logger.disabled = True
+# litellm.set_verbose = True
+
+# Based on the param support matrix below:
+# https://docs.litellm.ai/docs/completion/input
+# Some providers do not support the `n` parameter
+# and thus cannot generate multiple completions in one request
+unsupported_multiple_gen_providers = (
+    "openrouter/",
+    "claude",
+    "replicate/",
+    "bedrock",
+    "petals",
+    "palm/",
+    "together_ai/",
+    "text-bison",
+    "text-bison@001",
+    "chat-bison",
+    "chat-bison@001",
+    "chat-bison-32k",
+    "code-bison",
+    "code-bison@001",
+    "code-gecko@001",
+    "code-gecko@latest",
+    "codechat-bison",
+    "codechat-bison@001",
+    "codechat-bison-32k",
+)
+
+
+class LiteLLMGenerator(Generator):
+    """Generator wrapper using LiteLLM to allow access to different
+    providers using the OpenAI API format.
+    """
+
+    supports_multiple_generations = True
+    generator_family_name = "LiteLLM"
+
+    temperature = 0.7
+    top_p = 1.0
+    frequency_penalty = 0.0
+    presence_penalty = 0.0
+    stop = ["#", ";"]
+
+    def __init__(self, name: str, generations: int = 10):
+        self.name = name
+        self.fullname = f"LiteLLM {self.name}"
+        self.generations = generations
+        self.api_base = None
+        self.api_key = None
+        self.provider = None
+        self.supports_multiple_generations = not any(
+            self.name.startswith(provider)
+            for provider in unsupported_multiple_gen_providers
+        )
+
+        super().__init__(name, generations=generations)
+
+        if "litellm.LiteLLMGenerator" in _config.plugins.generators:
```

Comment:
Should this have have an `else` condition to raise an exception?  If this value is not found the generator is not usable.

Reply:
Yes